src/sandbox.ml

```diff
@@ -1,31 +1,68 @@
 open Import
 
-(* Terminology: - Sandbox: represents supported sandboxes with Global as the
-   fallback - project_root is different from Package_manager root (Eg. Opam
-   (Path.of_string "/foo/bar")). Project root is the directory where manifest
-   file (opam/esy.json/package.json) was found. Package_manager root is the
-   directory that contains the manifest file responsible for setting up the
-   sandbox - the two are same for Esy and Opam project but different for
-   bucklescript. Bucklescript projects have this manifest file abstracted away
-   from the user (at least at the moment) - Manifest: abstracts functions
-   handling manifest files of the supported package managers *)
+module Package = struct
+  type t =
```

Comment:
This type feels wrong to me for similar reasons as some other types that were removed in this PR. The reason why it feels wrong is that for either esy or opam, we always know there's kind of package we're dealing with in the tree view. It is never the case the case we're actually dealing with a collection of mixed packages. That suggests that we should use   a form of polymorphism in the sandbox treeview that reflects this. Either functors or type indexed values seem to fit the task much better.

Reply:
I agree the type does not modelize reality perfectly, however, I don't think it's a huge problem: the variants are not exposed in the interface, and the only way to get a list of packages is through `Sandbox.packages` and `Sandbox.root_packages`. 

FWIW, I tried removing the `Package` module and use functors in `Treeview_sandbox`. The problem is that we need the sandbox to instantiate the functor, and we have to get the sandbox in `getItems`, otherwise refreshing the treeview will not work as the function will use an old sandbox. Instantiating the functor in `getItems` is not possible, however, as the type of the functor instance would leak from the scope of the function (it won't compile).

In any case, having a `Package` module in `Sandbox` was intentional (a sandbox package that abstracts the package manager seemed like a good design), so I tried to use type-indexed value, as you suggested, with the following representation:

```ocaml
module Impl = struct
  type 'a t =
    { name : 'a -> string
    ; version : 'a -> string
    ; documentation : 'a -> string option
    ; synopsis : 'a -> string option
    ; has_dependencies : 'a -> bool
    ; dependencies : 'a -> ('a list, string) result Promise.t
    ; remove : string list -> (unit, string) result Promise.t
    ; install : string list -> (unit, string) result Promise.t
    ; upgrade : unit -> (unit, string) result Promise.t
    ; root_packages : unit -> ('a list, string) result Promise.t
    }
end

type 'a t =
  { pkg : 'a
  ; impl : 'a Impl.t
  }
```

But then functions such as `val packages : t -> (Package.t list, string) result Promise.t` can't be implemented.

Considering that the possibility of having a mixed list of packages is avoided by not exposing the type, I am not sure exploring other potential solutions is really worth it. Happy to try other solutions you might think of if you disagree though 🙂

Reply:
> But then functions such as val packages : t -> (Package.t list, string) result Promise.t can't be implemented.

The signature would need to be `val packages : 'a t -> ('a list, string) result Promise.t`. We'd have to push the type indexed value further.

Alright, then let's leave it as is for now. Just add a comment in the type describing the current caveats. We'll try to fix it another time.

Reply:
Done, I also pushed my explorations here:

- Functorized tree view: https://github.com/tmattio/vscode-ocaml-platform/commit/8011335d3c79b7a267122e7d62fe0dd13ae5bb67
- type-indexed values for sandbox packages: https://github.com/tmattio/vscode-ocaml-platform/commit/11c1dcdb36080427cb9b81c618ddc35e759298c3